components/feature/syncedtabs/src/main/java/mozilla/components/feature/syncedtabs/storage/SyncedTabsStorage.kt

```diff
@@ -27,26 +33,34 @@ import mozilla.components.support.ktx.kotlinx.coroutines.flow.ifChanged
 class SyncedTabsStorage(
     private val accountManager: FxaAccountManager,
     private val store: BrowserStore,
-    private val tabsStorage: RemoteTabsStorage = RemoteTabsStorage()
+    private val tabsStorage: RemoteTabsStorage = RemoteTabsStorage(),
+    private val debounceMillis: Long = 1000L,
 ) : SyncedTabsProvider {
     private var scope: CoroutineScope? = null
 
     /**
      * Start listening to browser store changes.
      */
+    @OptIn(FlowPreview::class)
     fun start() {
         scope = store.flowScoped { flow ->
-            flow.ifChanged { it.tabs }.map { state ->
+            flow.ifChanged { it.toSyncComponents() }.map { state ->
                 // TO-DO: https://github.com/mozilla-mobile/android-components/issues/5179
                 val iconUrl = null
-                state.tabs.filter { !it.content.private }.map { tab ->
+                state.tabs.filter { !it.content.private && !it.content.loading }.map { tab ->
                     // TO-DO: https://github.com/mozilla-mobile/android-components/issues/1340
                     val history = listOf(TabEntry(tab.content.title, tab.content.url, iconUrl))
                     Tab(history, 0, tab.lastAccess)
                 }
-            }.collect { tabs ->
-                tabsStorage.store(tabs)
             }
+                .debounce(debounceMillis)
+                .collect { tabs ->
+                    tabsStorage.store(tabs)
+                    accountManager.syncNow(
+                        reason = SyncReason.User,
+                        customEngineSubset = listOf(SyncEngine.Tabs)
+                    )
+                }
```

Comment:
Having there does mean this is no longer a "storage" later alone any more and it does other things, which you've identified already.

How about we try and create a delegate that would perform post-collect tasks. We have a SyncTabsController (update it to with a syncNow `customEngineSubset`), so we could sort of use that here too.

```kotlin
class SyncedTabsStorage(
    private val accountManager: FxaAccountManager,
    private val store: BrowserStore,
    private val tabsStorage: RemoteTabsStorage = RemoteTabsStorage(),
    // This could be an interface too that calls a controller.syncNow() instead of an anonymous lambda.
    private val onStoreComplete: () -> Unit = { accountManager.syncNow() } 
)
```

```suggestion
                .collect { tabs ->
                    tabsStorage.store(tabs)
                    onStoreComplete()
                }
```

Reply:
Just to clarify, are you suggesting we inject the `SyncTabsController` into the storage and use that for the default `onStoreComplete` delegate, or just to use the suggested code change above?

Reply:
I'll add the custom sync engine subset to the `SyncTabsController` implementation as well